File: rmw_int2dds_validation/validation/qos/deadline/rclcpp/qos_deadline_sub.cpp
// Copyright 2026 Int2DDS Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/subscription_options.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

namespace
{

class DeadlineSubscriberCpp : public rclcpp::Node
{
public:
  DeadlineSubscriberCpp(
    const std::string & topic,
    double deadline_sec,
    double timeout_sec,
    int expect_at_least,
    bool expect_none,
    bool require_deadline_event,
    bool require_incompatible_event)
  : Node("qos_deadline_sub_cpp",
      rclcpp::NodeOptions().start_parameter_services(false).enable_rosout(false)),
    topic_(topic),
    expect_at_least_(expect_at_least),
    expect_none_(expect_none),
    require_deadline_event_(require_deadline_event),
    require_incompatible_event_(require_incompatible_event)
  {
    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.reliability(rclcpp::ReliabilityPolicy::Reliable);
    qos.durability(rclcpp::DurabilityPolicy::Volatile);
    qos.deadline(std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::duration<double>(
        deadline_sec)));

    rclcpp::SubscriptionOptions options;
    if (require_deadline_event) {
      options.event_callbacks.deadline_callback =
        [this](rclcpp::QOSDeadlineRequestedInfo & info) {
          deadline_events_.emplace_back(info.total_count, info.total_count_change);
          RCLCPP_INFO(
            get_logger(),
            "subscription deadline event: total_count=%d, total_count_change=%d",
            info.total_count,
            info.total_count_change);
        };
    }
    if (require_incompatible_event) {
      options.event_callbacks.incompatible_qos_callback =
        [this](rclcpp::QOSRequestedIncompatibleQoSInfo & info) {
          incompatible_events_.push_back(
            std::make_tuple(info.total_count, info.total_count_change, info.last_policy_kind));
          RCLCPP_INFO(
            get_logger(),
            "subscription incompatible qos event: total_count=%d, total_count_change=%d,"
            "last_policy_kind=%d",
            info.total_count,
            info.total_count_change,
            info.last_policy_kind);
        };
    }

    subscription_ = create_subscription<std_msgs::msg::String>(
      topic,
      qos,
      [this](const std_msgs::msg::String::SharedPtr msg) {
        received_messages_.push_back(msg->data);
        RCLCPP_INFO(get_logger(), "subscriber heard: %s", msg->data.c_str());
      },
      options);
    observation_timer_ = create_wall_timer(500ms, [this]() {
          RCLCPP_INFO(get_logger(), "publisher_count observation: %zu", count_publishers(topic_));
    });
    finish_timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::duration<double>(
        timeout_sec)),
      [this]() {
        finish_timer_->cancel();
        observation_timer_->cancel();
        done_ = true;
      });
    RCLCPP_INFO(
      get_logger(), "subscriber ready: topic=%s, deadline=%.3fs", topic.c_str(), deadline_sec);
  }

  bool done() const {return done_;}

  bool evaluate()
  {
    bool ok = true;
    if (expect_none_) {
      ok = ok && received_messages_.empty();
    } else {
      ok = ok && static_cast<int>(received_messages_.size()) >= expect_at_least_;
    }
    if (require_deadline_event_) {
      ok = ok && !deadline_events_.empty();
    }
    if (require_incompatible_event_) {
      ok = ok && !incompatible_events_.empty();
    }

    if (ok) {
      if (expect_none_) {
        RCLCPP_INFO(get_logger(), "deadline subscriber ok: received nothing");
      } else {
        RCLCPP_INFO(get_logger(), "deadline subscriber ok: received_count=%zu",
            received_messages_.size());
      }
    } else {
      RCLCPP_ERROR(
        get_logger(),
        "deadline subscriber failed: received_count=%zu, deadline_events=%zu,"
        "incompatible_events=%zu",
        received_messages_.size(), deadline_events_.size(), incompatible_events_.size());
    }
    return ok;
  }

private:
  std::string topic_;
  int expect_at_least_;
  bool expect_none_;
  bool require_deadline_event_;
  bool require_incompatible_event_;
  bool done_{false};
  std::vector<std::string> received_messages_;
  std::vector<std::pair<int, int>> deadline_events_;
  std::vector<std::tuple<int, int, int>> incompatible_events_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr observation_timer_;
  rclcpp::TimerBase::SharedPtr finish_timer_;
};

}  // namespace

int main(int argc, char ** argv)
{
  std::string topic;
  double deadline_sec = 0.2;
  double timeout_sec = 8.0;
  int expect_at_least = 0;
  bool expect_none = false;
  bool require_deadline_event = false;
  bool require_incompatible_event = false;

  for (int i = 1; i < argc; ++i) {
    const std::string arg(argv[i]);
    auto next_value = [&](const char * name) -> std::string {
        if (i + 1 >= argc) {
          throw std::runtime_error(std::string("missing value for ") + name);
        }
        return argv[++i];
      };
    if (arg == "--topic") {
      topic = next_value("--topic");
    } else if (arg == "--deadline-sec") {
      deadline_sec = std::stod(next_value("--deadline-sec"));
    } else if (arg == "--timeout") {
      timeout_sec = std::stod(next_value("--timeout"));
    } else if (arg == "--expect-at-least") {
      expect_at_least = std::stoi(next_value("--expect-at-least"));
    } else if (arg == "--expect-none") {
      expect_none = true;
    } else if (arg == "--require-deadline-event") {
      require_deadline_event = true;
    } else if (arg == "--require-incompatible-event") {
      require_incompatible_event = true;
    } else {
      throw std::runtime_error("unknown argument: " + arg);
    }
  }

  if (topic.empty()) {
    throw std::runtime_error("--topic is required");
  }
  if (expect_none && expect_at_least > 0) {
    throw std::runtime_error("--expect-none and --expect-at-least cannot be used together");
  }

  rclcpp::init(argc, argv);
  auto node = std::make_shared<DeadlineSubscriberCpp>(
    topic, deadline_sec, timeout_sec, expect_at_least, expect_none, require_deadline_event,
    require_incompatible_event);
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  while (rclcpp::ok() && !node->done()) {
    executor.spin_once(100ms);
  }
  auto ok = node->evaluate();
  executor.remove_node(node);
  rclcpp::shutdown();
  return ok ? 0 : 1;
}

File: validation/qos/deadline/rclpy/qos_deadline_sub.py
#!/usr/bin/env python3

# Copyright 2026 Int2DDS Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import sys

import rclpy
from rclpy.duration import Duration
try:
    from rclpy.event_handler import SubscriptionEventCallbacks
except ModuleNotFoundError:
    from rclpy.qos_event import SubscriptionEventCallbacks
from rclpy.executors import SingleThreadedExecutor
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile, ReliabilityPolicy
from std_msgs.msg import String


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument('--topic', required=True)
    parser.add_argument('--deadline-sec', type=float, required=True)
    parser.add_argument('--timeout', type=float, default=8.0)
    parser.add_argument('--expect-at-least', type=int, default=0)
    parser.add_argument('--expect-none', action='store_true')
    parser.add_argument('--require-deadline-event', action='store_true')
    parser.add_argument('--require-incompatible-event', action='store_true')
    return parser.parse_args()


class DeadlineSubscriber(Node):
    def __init__(self, args: argparse.Namespace) -> None:
        super().__init__('qos_deadline_sub', start_parameter_services=False, enable_rosout=False)
        self.args = args
        self.done = False
        self.received_messages = []
        self.deadline_events = []
        self.incompatible_events = []

        qos = QoSProfile(
            history=HistoryPolicy.KEEP_LAST,
            depth=10,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            deadline=Duration(seconds=args.deadline_sec),
        )
        callbacks = SubscriptionEventCallbacks(
            deadline=self._deadline_callback if args.require_deadline_event else None,
            incompatible_qos=(
                self._incompatible_callback if args.require_incompatible_event else None),
        )
        self.subscription = self.create_subscription(
            String, args.topic, self._callback, qos, event_callbacks=callbacks
        )
        self.observation_timer = self.create_timer(0.5, self._observe)
        self.finish_timer = self.create_timer(max(args.timeout, 0.1), self._finish)
        self.get_logger().info(
            'subscriber ready: '
            f'topic={args.topic}, deadline={args.deadline_sec}s'
        )

    def _callback(self, msg: String) -> None:
        self.received_messages.append(msg.data)
        self.get_logger().info(f'subscriber heard: {msg.data}')

    def _deadline_callback(self, event) -> None:
        self.deadline_events.append((event.total_count, event.total_count_change))
        self.get_logger().info(
            'subscription deadline event: '
            f'total_count={event.total_count}, total_count_change={event.total_count_change}'
        )

    def _incompatible_callback(self, event) -> None:
        self.incompatible_events.append(
            (event.total_count, event.total_count_change, event.last_policy_kind)
        )
        self.get_logger().info(
            'subscription incompatible qos event: '
            f'total_count={event.total_count}, total_count_change={event.total_count_change}, '
            f'last_policy_kind={event.last_policy_kind}'
        )

    def _observe(self) -> None:
        self.get_logger().info(
            f'publisher_count observation: {self.count_publishers(self.args.topic)}'
        )

    def _finish(self) -> None:
        self.finish_timer.cancel()
        self.observation_timer.cancel()
        self.done = True

    def evaluate(self) -> bool:
        ok = True
        if self.args.expect_none:
            ok = ok and len(self.received_messages) == 0
        else:
            ok = ok and len(self.received_messages) >= self.args.expect_at_least
        if self.args.require_deadline_event:
            ok = ok and len(self.deadline_events) >= 1
        if self.args.require_incompatible_event:
            ok = ok and len(self.incompatible_events) >= 1

        if ok:
            if self.args.expect_none:
                self.get_logger().info('deadline subscriber ok: received nothing')
            else:
                self.get_logger().info(
                    f'deadline subscriber ok: received_count={len(self.received_messages)}'
                )
        else:
            self.get_logger().error(
                'deadline subscriber failed: '
                f'received={self.received_messages}, '
                f'deadline_events={self.deadline_events}, '
                f'incompatible_events={self.incompatible_events}'
            )
        return ok


def main() -> int:
    args = parse_args()
    if args.expect_none and args.expect_at_least > 0:
        raise RuntimeError('--expect-none and --expect-at-least cannot be used together')
    rclpy.init()
    node = DeadlineSubscriber(args)
    executor = SingleThreadedExecutor()
    executor.add_node(node)
    try:
        while rclpy.ok() and not node.done:
            executor.spin_once(timeout_sec=0.1)
        ok = node.evaluate()
    finally:
        executor.remove_node(node)
        node.destroy_node()
        executor.shutdown(timeout_sec=1.0)
        rclpy.shutdown()
    return 0 if ok else 1


if __name__ == '__main__':
    sys.exit(main())
